src/dynamic_query.rs

```diff
@@ -0,0 +1,130 @@
+use core::{alloc::Layout, any::TypeId, ptr::NonNull};
+
+use crate::{entities::EntityMeta, Archetype, Entity};
+
+/// The component types accessed by a dynamic query.
+#[derive(Debug, Copy, Clone)]
+pub struct DynamicQueryTypes<'a> {
+    pub(crate) read_types: &'a [TypeId],
+    pub(crate) write_types: &'a [TypeId],
+}
+
+impl<'a> DynamicQueryTypes<'a> {
+    /// Creates a dynamic query that reads component types in `read_types`
+    /// and writes component types in `write_types`.
+    pub fn new(read_types: &'a [TypeId], write_types: &'a [TypeId]) -> Self {
+        Self {
+            read_types,
+            write_types,
+        }
+    }
+
+    /// Gets the component types to be read by this query.
+    pub fn read_types(&self) -> &[TypeId] {
+        self.read_types
+    }
+
+    /// Gets the component types to be accessed mutably by this query.
+    pub fn write_types(&self) -> &[TypeId] {
+        self.write_types
+    }
+}
+
+/// A pointer to a slice of component data from a [`DynamicQuery`].
+pub struct ComponentSlice {
+    typ: TypeId,
+    ptr: NonNull<u8>,
+    len: usize,
+    component_layout: Layout,
+}
+
+impl ComponentSlice {
```

Comment:
There should be a safe way to get mutable access to the `write_types` of the query.